controllers/network/ipset_controller.go

```diff
@@ -0,0 +1,394 @@
+			// and recalculate it based on the state of the rest of the conditions
+			instance.Status.Conditions.Set(
+				instance.Status.Conditions.Mirror(condition.ReadyCondition))
+		}
+
+		err := helper.PatchInstance(ctx, instance)
+		if err != nil {
+			_err = err
+			return
+		}
+	}()
+
+	// If we're not deleting this and the service object doesn't have our finalizer, add it.
+	if instance.DeletionTimestamp.IsZero() && controllerutil.AddFinalizer(instance, helper.GetFinalizer()) {
+		return ctrl.Result{}, err
+	}
+
+	// initialize status
+	if instance.Status.Conditions == nil {
+		instance.Status.Conditions = condition.Conditions{}
+
+		cl := condition.CreateList(
+			condition.UnknownCondition(condition.ServiceConfigReadyCondition, condition.InitReason, condition.ServiceConfigReadyInitMessage),
+		)
+
+		instance.Status.Conditions.Init(&cl)
+
+		// Register overall status immediately to have an early feedback e.g. in the cli
+		return ctrl.Result{}, nil
+	}
+
+	instance.Status.Reservation = []networkv1.IPAddress{}
+
+	// Handle service delete
+	if !instance.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, instance, helper)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcileNormal(ctx, instance, helper)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *IPSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&networkv1.IPSet{}).
+		Owns(&networkv1.Reservation{}).
+		Complete(r)
+}
+
+func (r *IPSetReconciler) reconcileDelete(ctx context.Context, instance *networkv1.IPSet, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service delete")
+
+	// get reservation
+	res, err := r.getReservation(ctx, instance)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if err := r.delFinalizerOnReservation(ctx, res, helper); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Service is deleted so remove the finalizer.
+	controllerutil.RemoveFinalizer(instance, helper.GetFinalizer())
+	r.Log.Info("Reconciled Service delete successfully")
+
+	return ctrl.Result{}, nil
+}
+
+func (r *IPSetReconciler) reconcileNormal(ctx context.Context, instance *networkv1.IPSet, helper *helper.Helper) (ctrl.Result, error) {
+	r.Log.Info("Reconciling Service")
+
+	instance.Status.Conditions.MarkTrue(condition.ServiceConfigReadyCondition, condition.InputReadyMessage)
+
+	r.Log.Info("get all Reservations")
+	// get list of Reservation objects in the namespace
+	reservations := &networkv1.ReservationList{}
+	opts := &client.ListOptions{
+		Namespace: instance.Namespace,
+	}
+	err := r.List(ctx, reservations, opts)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// reservations map map[string]Reservation with reservation name as index
+	resMapNameIdx := make(map[string]networkv1.Reservation)
+	// map with all ip -> hostname references
+	resMapIPIdx := make(map[string]string)
```

Comment:
While convenient, this map looks like a potential hot-spot at scale?
If I undestrand it correctly it will iterate over all IPs already reserved in all networks for each individual new allocation?

I think we can use the `slices.IndexFunc` in `pkg/ipam/funcs.go`'s `fixedIPExists()` and `iterateForAssignment()`. If we can operate directly against `ReservationList`  there instead I believe we would avoid iterating trough everything every time?

Like this:
```
	f := func(c Reservation) bool {
		for _, v := range c.Spec.Reservations {
			if v.Address == nextip {
				return true
			}
		}
		return false
	}
	idx := slices.IndexFunc(reservations.Items, f)
	if  (idx >= 0 && reservations.Items[idx].Spec.Hostname != a.Hostname) {
		fmt.Println(fmt.Sprintf("IP: %s already reseverd for host: %s", nextip, a.Hostname))
	}

```

Or, if we can add NetName name to `AssignIPDetails` we could do:
```
	f := func(c Reservation) bool {
		return c.Spec.Reservations[a.NetName].Address == nextip
	}
	idx := slices.IndexFunc(reservations.Items, f)
	if (idx >= 0 && reservations.Items[idx].Spec.Hostname != a.Hostname) {
		fmt.Println(fmt.Sprintf("IP: %s already reseverd for host: %s", nextip, a.Hostname))
	}
```